From a119fa4c7931d0d9d6650fd91ab0e6ea240132a9 Mon Sep 17 00:00:00 2001
From: Jianhua Lu <lujianhua000@gmail.com>
Date: Wed, 29 Mar 2023 19:38:33 +0800
Subject: [PATCH] Asoc: wm_adsp: Add prefix support

Signed-off-by: Jiali Chen <chenjiali@radxa.com>
---
 sound/soc/codecs/wm_adsp.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/sound/soc/codecs/wm_adsp.c b/sound/soc/codecs/wm_adsp.c
index ca630c994..6338944b5 100644
--- a/sound/soc/codecs/wm_adsp.c
+++ b/sound/soc/codecs/wm_adsp.c
@@ -766,6 +766,10 @@ static int wm_adsp_request_firmware_file(struct wm_adsp *dsp,
 		fw->filename = kasprintf(GFP_KERNEL, "%s%s-%s-%s-%s.%s", dir, dsp->part,
 					 fwf, wm_adsp_fw[dsp->fw].file, system_name,
 					 filetype);
+	else if (asoc_component_prefix)
+		fw->filename = kasprintf(GFP_KERNEL, "%s%s-%s-%s-%s.%s", dir, dsp->part,
+					 fwf, wm_adsp_fw[dsp->fw].file, asoc_component_prefix,
+					 filetype);
 	else
 		fw->filename = kasprintf(GFP_KERNEL, "%s%s-%s-%s.%s", dir, dsp->part, fwf,
 					 wm_adsp_fw[dsp->fw].file, filetype);
@@ -854,6 +858,23 @@ VISIBLE_IF_KUNIT int wm_adsp_request_firmware_files(struct wm_adsp *dsp,
 
 		if (fw->wmfw.firmware || (dsp->wmfw_optional && fw->coeff.firmware))
 			return 0;
+	} else if (suffix) {
+		ret = wm_adsp_request_firmware_file(dsp, &fw->wmfw,
+						    cirrus_dir, NULL,
+						    NULL, "wmfw");
+		if (ret < 0)
+			goto err;
+
+		if (fw->wmfw.firmware) {
+			adsp_dbg(dsp, "Found '%s'\n", fw->wmfw.filename);
+			ret = wm_adsp_request_firmware_file(dsp, &fw->coeff,
+							    cirrus_dir, NULL,
+							    suffix, "bin");
+			if (ret < 0)
+				goto err;
+
+			return 0;
+		}
 	}
 
 	/* Check legacy location */
-- 
2.54.0

